solver/src/main/java/org/chocosolver/solver/constraints/binary/PropSquare.java

```diff
@@ -117,57 +108,95 @@ protected void updateLowerBoundofX() throws ContradictionException {
 
     protected void updateUpperBoundofX() throws ContradictionException {
         vars[0].updateUpperBound(Math.max(sqr(vars[1].getLB()), sqr(vars[1].getUB())), this);
-
     }
 
     protected boolean updateHolesinX() throws ContradictionException {
+        if (!vars[0].hasEnumeratedDomain()) {
+            return false;
+        }
+        boolean impact = false;
+        boolean isFirstPropagation = model.getSolver().getNodeCount() == 0;
+        if (isFirstPropagation) {
+            // check perfect squares once and for all
+            int ub = vars[0].getUB();
+            vrms.clear();
+            vrms.setOffset(vars[0].getLB());
+            for (int value = vars[0].getLB(); value <= ub; value = vars[0].nextValue(value)) {
+                if (!MathUtils.isPerfectSquare(value)) {
+                    vrms.add(value);
+                }
+            }
+            impact = vars[0].removeValues(vrms, this);
+        }
+
         // remove intervals to deal with consecutive value removal and upper bound modification
         if (bothEnum) {
             int ub = vars[0].getUB();
             vrms.clear();
             vrms.setOffset(vars[0].getLB());
             for (int value = vars[0].getLB(); value <= ub; value = vars[0].nextValue(value)) {
-                if (!(MathUtils.isPerfectSquare(value) &&
-                        (vars[1].contains(floor_sqrt(value)) || vars[1].contains(-floor_sqrt(value))))) {
+                int sqrt = floor_sqrt(value);
+                if (!vars[1].contains(sqrt) && !vars[1].contains(-sqrt)) {
                     vrms.add(value);
                 }
             }
-            return vars[0].removeValues(vrms, this);
-        } else if (vars[0].hasEnumeratedDomain()) {
+            impact |= vars[0].removeValues(vrms, this);
+        } else {
```

Comment:
The "else" might be only if vars[1].hasEnumeratedDomain(), doesn't it ? If vars[0] has an enumerated domain (and therefore vars[1] has not), this part of code will never do anything (since there are no two integers i and j such that i^2 == j^2 - 1, except 0 and 1) which is not done in updateLowerBoundofX() or updateUpperBoundofX().

Moreover, inside the while loops, value should be updated with the next (or previous) perfect square value instead of vars[0].nextValue(value) (or vars[0].previousValue(value)) as these values won't have been removed at root node if vars[0] has not an enumerated domain. By the way, such methods should be used directly in updateLowerBoundofX() and updateUpperBoundofX().

That said, with these corrections, I don't see a situation where this "else" code is doing something (especially with the statement returning false if vars[0] has not an enumerated domain).

Reply:
The else had a vars[0].hasEnumeratedDomain() condition that is now at the beginning of the methode.

Actually, I was thinking that maybe the else should be completely removed. I am not sure there is more to filter in such case. Isn't this filtering already ensured by the setBounds() ?

Reply:
If there is more to filter, could you please provide an exemple / test case in SquareTest ?

Reply:
It was my point (sorry if it was not clear). I think the else is useless whenever vars[1] has not an enumerated domain, as the setBounds() method should already have filtered the bounds.

However, there is one thing that should be added to the updateLowerBoundofX() and updateUpperBoundofX() methods: a check that the new bound is a perfect square (which is not the case if both variables haven't enumerated domains and y.LB <= 0 <= y.UB).